### ISSUE TITLE
Document usage and behavior of web examples

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5612]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)| @argonemyth reported|
|-|-|
|Trac ID|trac#5612|
|Type|task|
|Created|2012-04-04 16:16:40Z|
|Branch|https://github.com/twisted/twisted/tree/web-example-improvements-5612|

The examples in doc/web/examples mostly fail to explain how they are used and what they accomplish.  Some documentation would make them more useful.

Attachments:

* [web-examples.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f75/f75c77b610903e69c983cf7f8a3918e779f691cd/f7fd4e8d8c39637fc75474936e0bb7388c464a40.patch) (18574 bytes) - added by argonemyth on 2012-04-10 07:51:38Z - Patch for directory doc/web/examples
* [web-examples-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f75/f75c77b610903e69c983cf7f8a3918e779f691cd/c6511e5ba64f6bf4b446ede92dae265b811fb65a.patch) (18664 bytes) - added by argonemyth on 2012-04-10 12:21:19Z - 
* [web-examples-v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f75/f75c77b610903e69c983cf7f8a3918e779f691cd/e893e75e1ab95e2ae2e7dfe0532bb2d9fa97277e.patch) (25530 bytes) - added by argonemyth on 2012-04-16 11:17:53Z - 
* [web-examples-v4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f75/f75c77b610903e69c983cf7f8a3918e779f691cd/48cfc5f13fec14f6bc1d648a4c395cf3b3d7fef4.patch) (25262 bytes) - added by argonemyth on 2012-04-16 14:02:54Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5612 5612
type__task task
reporter__argonemyth argonemyth
priority__lowest lowest
milestone__ 
branch__branches_web_example_improvements_5612 branches/web-example-improvements-5612
branch_author__argonemyth__exarkun argonemyth, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__documentation documentation
time__1333556200000000 1333556200000000
changetime__1335109317000000 1335109317000000
version__None None
owner__exarkun exarkun
cc__jesstess cc__jknight cc__argonemyth
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)|@argonemyth set owner to @argonemyth<br>@argonemyth set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_4"></a>|@argonemyth set owner to @jesstess<br>@argonemyth set status to `new`|
|-|-|

I added comments to most of the examples & fixed some scripts that were no longer working properly. Please review :)

Here is a brief list of code changes:

* logging-proxy.py --> since the proxy doesn't support HTTPS, it threw out lots of errors when visiting a https site, I caught the exception to make the output a bit nicer

* xmlrpcclient.py --> deleted two examples - sites no longer exist

* soap.py  --> changed the echo example, as I got 'None' when giving the echo method a list

* lj.rpyp.py  --> changed RRS URL and added timeout

* report.rpy.py --> changed the way it gets host and port, the old method no longer works

* users.rpy.py  --> added try & except - registry.getComponent(distrib.UserDirectory) throws an exception

* rootscript.py --> made it a proper twisted web application

Note: simple.rtl doesn't work due to a bug in twisted.web.script.ResourceTemplate @ line 73, should change it to : from quixote.ptl import ptl_compile.

For more details, please check the web-examples.patch!

Fei

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @argonemyth|
|-|-|

Thanks for the patch.

 1. Some of the examples touched in this patch were updated recently (ie. google.py, lj.rpy.py, and getpage.py) and you should make your changes against the latest svn trunk. 
 1. In some of the examples (ie. dlpage.py) both a comment and docstring are used to describe the example, I would move both of them in a docstring instead.
 1. Every example needs a license header
 1. in `logging-proxy.py`:
   * the word `protocal` should be spelled `protocol` (same mistake is in proxy.py)
   * you add an urllib import but it's not being used 

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_6"></a>|@argonemyth set owner to @jesstess|
|-|-|

Thanks for reviewing! Can't believe I typed 'protocol' wrong, how embarrassing! Well, I just co'ed the latest svn trunk and modified the examples accordingly. Please review it again :)

Patch name: web-examples-v2.patch

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @argonemyth|
|-|-|

Thanks!

  1. When generating patches, please generate them from the top-level of the Twisted checkout.  That is, if you have Twisted checked out in "Twisted-trunk", run the "svn diff" command from "Twisted-trunk", not from "Twisted-trunk/doc/web".  This gives files in the patch the correct names and makes it easier to apply.
  1. Make sure to avoid having trailing whitespace on a line.  Eg, "Foo bar baz" not "Foo bar baz ".  If you use emacs, you can have it highlight trailing whitespace by customizing the trailing-whitespace face.
  1. I suppose aiming for proper punctuation everywhere is a good idea, as long as we're doing this. :)  eg, periods at the end of sentences, all that stuff.
  1. in `doc/web/examples/dlpage.py`, the string you added above the call to `downloadPage` should be a comment.  Strings are only docstrings if they are the first statement in a module, class, or function.  docstrings are special because they get attached to the object as the `__doc__` attribute.  Strings elsewhere are just no-op statements.  When they are exposition for readers, the information should be in comments instead.  Note this isn't a contradiction of what jesstess recommended in her review - any information it makes sense to present to a reader about the file as a whole should be in a module docstring, not split between a module docstring and a comment near the top of the module.
  1. `getpage.py` should be converted from starting with a comment explaining usage to starting with a docstring explaining usage.
  1. in `httpclient.py`, you added some docstrings like `""" foo """`.  Always write docstrings on at least three lines: `"""\nfoo\n"""`.
  1. As far as the new error handling in `logging-proxy.py` goes, I'm not sure a print there is too helpful.  The browser won't see the error, for example.  It might be better to leave the code as it was, perhaps filing a ticket to improve it to do real error reporting.
  1. There's a new "bare word except" in users.rpy.py.  This is bad practice in almost all Python code.  What problem is this aimed at solving?
  1. The ticket description says something about testing, but the patch doesn't attempt to address that.  I think that's a good idea.  We should eventually have unit tests for all our examples, but documentation is a good first step.  Can you file a new ticket for doing the unit testing, though?

Some of these comments apply to more than just the file I pointed out.  I didn't list all the files where the changes are needed, but please do update all of them.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_9"></a>|@argonemyth commented|
|-|-|

Exarkun,

Thanks for listing the guidelines for the documentation. I will submit a new patch soon!

For No.7 (logging-proxy.py), I added the error handling because it clears out the frustration for new Twisted users. My first response to the error was 'huh, the sample was broken!'. I didn't know it only supported http by default until I checked the source code...no document actually mentioned this. I think by adding the exception will at least tell them what's going on, and a blank page for google.com is somehow 'expected'.

As for the 'testing', what I meant was to 'run the script and see if it works'. I am not familiar how Twisted handles unit testing. Well, I guess it's a good idea to file a new ticket for this. I will do that once I know what to do. :)

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_10"></a>|@argonemyth removed owner|
|-|-|

Third patch uploaded, please review, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @argonemyth|
|-|-|

Thanks.

 1. doc/web/examples/fortune.rpy.py:
   1. `Change` should be lowercase `change` in "# Link your fortune program to /usr/games or Change the path"
 1. doc/web/examples/dlpage.py:
   1.  `save` should be `saves`
 1. doc/web/examples/webguard.py:
   1. In `This example shows how to make simple web authentication` the word `make` should be `implement`. The sentence also needs a period at the end.
   1.  There's a trailing space behind `..(joe='blow')]` that needs to go
 1. doc/web/examples/advogato.py:
   1. sentences need to end with a period.
  1. doc/web/examples/xmlrpc.py:
   1. The statement about xmlrpc in Python 2.2 can be removed
  1. doc/web/examples/rootscript.py:
   1. the reference to the path of /etc/hosts can be removed, because win32 uses a different path. simply state the hosts file.
 
In general: all sentences need to start with a capital and end with a period.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_12"></a>|@argonemyth removed owner|
|-|-|

Thanks for reviewing! I just uploaded a new patch. Sorry that I didn't pay enough attention on the grammar!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [34248]) Branching to 'web-example-improvements-5612'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [34249]) Apply web-examples-v4.patch 

refs [#5612](https://github.com/twisted/twisted/issues/5612)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [34250]) Some cleanups

  * Remove all trailing whitespace
  * Normally the "usage" formatting
  * Try to make sure each example starts with an explanation of what it is an example of
  * Reflow lines to ~80 columns
  * Move stdlib imports above Twisted imports
  * Clean up some spelling errors and misc punctuation issues
  * Remove unneeded `__future__.nested_scopes` import

refs [#5612](https://github.com/twisted/twisted/issues/5612)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [34251]) whitespace cleanup and remove vhost.rpy.py from the index.  Also delete the vhost.rpy.py example.  It is a duplicate of web.py and the run instructions are insanely complex.

refs [#5612](https://github.com/twisted/twisted/issues/5612)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

I checked the patch in to a branch and made some further changes to it, linked in comments above.  I guess I won't try to be the reviewer on this, now.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Thanks for the additional fixes. Please merge after the doc builder succeeds.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Thanks.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/web-example-improvements-5612) look good, merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [34258]) Merge web-example-improvements-5612

Author: argonemyth, exarkun
Reviewer: exarkun, thijs
Fixes: [#5612](https://github.com/twisted/twisted/issues/5612)

Add and improve documentation for the Twisted Web examples.  Also remove one redundant
vhost monster example.
